contrib/clients/xim/quantizim.man
.TH QUANTIZIM 1 "February 3,  1989"
.UC 5
.SH NAME
quantizim \- quantize a 24 bit/pixel image to 256 colors or less.
.SH SYNOPSIS
.B quantizim
[-in 
.I file
] [-out
.I file
] [-colors
.I n] [-nodes
.I n
]
.SH DESCRIPTION
.I Quantizim
is a program for quantizing 24 bit/pixel images down to 8 bits or less.
It does a fairly good job creating a pixmap and colormap from a three channel RG&B data.
At the heart of this program is the
.I alcols
code written at the USC Information Sciences Institute.
It uses a tree sorting algorithm for selecting the best 256 (or less) unique colors from over 16 million.
.SH OPTIONS
.TP "\w'\f3\-m\f1name 'u"
.B -in \fIfile\fP
This option allows the user to explicitly specify the 24 bit/pixel file on the command line.  The default is to read input from standard in.
.TP "\w'\f3\-m\f1name 'u"
.B -out \fIfile\fP
This option allows the user to explicitly specify the 8 bit/pixel file on the command line.
The default is to write output to standard out.
.TP "\w'\f3\-m\f1name 'u"
.B -colors \fInum\fP
This allows one to specify the maximum number of colors to quantize down to.
The program may return less than the maximum.
The default is 256 colors.x.
.TP "\w'\f3\-m\f1name 'u"
.B -nodes \fInum\fP
The ratio of pixels per node to begin the tree sort with.
The default is 50.
A smaller ratio (more nodes) gives more even shading and uses more memory.
Less nodes causes more banding but generally less quantization error.
.PP
.I Note:
All of the command line options can be abbreviated to their shortest unique spelling.
.br
ex: %  quantizim -in colord24.xim -out color.xim -co 192 -no 10

.SH "SEE ALSO"
xim(n)
smoothim(n)
.SH BUGS
- This program takes a lot of memory.
Large images around 640x480 may not work on all machines.
.br
- The optimum ratio of nodes to image size is not clear.
.br
- I wish I knew better how this works.
.SH COPYRIGHTS
Copyright (c) 1988, University of Southern California
.br
Copyright (c) 1989, Massachusetts Institute of Technology.
.SH AUTHORS
Paul Raveling (vaxa.isi.edu) - alcols.c
.PP
Philip R. Thompson (phils@athena.mit.edu) - quantizim.c
.br
Computer Resource Laboratory,  
Dept. of Architecture and Planning,  
.br
Massachusetts Institute of Technology,  
Cambridge, MA  02139



contrib/clients/xim/quantizim.c
/**********************************************************************/
/* file:    quantizim.c                                               */
/* purpose: To quantize a 2**24 bit image down to 2**8 (256 or less)  */
/*          with an ImageHeader                                       */
/* author:  Philip Thompson, 1988                                     */
/* Copyright (c) 1989, Massachusetts Institute of Technology         */
/*                Philip R. Thompson   (phils@athena.mit.edu)         */
/*                Computer Resource Laboratory (CRL)                  */
/*                Dept. of Architecture and Planning                  */
/*                M.I.T., Rm 9-526                                    */
/*                Cambridge, MA  02139                                */
/**********************************************************************/
#ifndef lint
static char  quantizim_rcsid[] =
    "$Header: /mit/phils/X/RCS/quantizim.c,v 1.14 89/11/09 17:27:23 phils Exp Locker: phils $";
#endif

#include <stdio.h>
#include <strings.h>
#include "XimHeader.h"

#define OCTREE  1
#define DITHER  2
#define MEDIAN  3
                          
char *progName;
char *malloc(), *calloc();
extern debug_flag;

main(argc, argv)
int argc;
char **argv;
{
    register  i;
    XimImage xim;
    int ncolors = 256, enhance = 0, type = OCTREE;
    FILE *out_fp = stdout, *in_fp = stdin, *popen(), *fopen();
    int  pclose(), fclose(), (*closefunc)() = fclose;
    Color  *colors;
    extern int  COLTNODES, ditherFactor;
    int map_flag = 0, dither_flag = 1;

    progName = argv[0];
    for (i=1; i < argc; i++) {
        char *ptr = NULL;
        if (strncmp(argv[i], "-i", 2) == 0) {
            ptr = rindex(argv[++i], '.');
            if (ptr && strcmp(ptr, ".Z") == 0) {   /* compressed file */
                char popen_cmd[80];
                (void)strcpy(popen_cmd, "zcat ");
                (void)strcat(popen_cmd, argv[i]);
                in_fp = popen(popen_cmd, "r");
                closefunc = pclose;
            } else
                in_fp = fopen(argv[i], "r");
            if (in_fp == NULL)
                XimError("Can't open input file: %s ", argv[i]);
            continue;
        }
        if (strncmp(argv[i], "-o", 2) == 0) {
            if ((out_fp = fopen(argv[++i], "w")) == NULL)
                XimError("Can't open %s\n", argv[i]);
            continue;
        }
        if (strncmp(argv[i], "-co", 3) == 0) {
            ncolors = atoi(argv[++i]);
            continue;
        }
        if (strncmp(argv[i], "-di", 3) == 0) {
            type = DITHER;
            if ((ptr=rindex(argv[i],'=')) != NULL)
                ditherFactor = atoi(ptr+1);
            else
                ditherFactor = 6;
            dither_flag = !dither_flag; /* toggle median dithering */
            continue;
        }
        if (strncmp(argv[i], "-me", 3) == 0) {
            type = MEDIAN;
            continue;
        }
        if (strncmp(argv[i], "-cm", 3) == 0) {
            map_flag = 1;
            type = MEDIAN;
            continue;
        }
        if (strncmp(argv[i], "-no", 3) == 0) {
            COLTNODES = atoi(argv[++i]);
            continue;
        }
        if (strncmp(argv[i], "-en", 3) == 0) {
            if ((ptr=rindex(argv[i],'=')) != NULL)
                enhance = atoi(ptr+1);
            else
                enhance = 9;
            continue;
        }
        if (strncmp(argv[i], "-de", 3) == 0) {
            debug_flag = 1;
            continue;
        }
        Usage();
    }

    ReadXim(in_fp, &xim);
    (void)(*closefunc)(in_fp);
    if (xim.nchannels != 3 || xim.bits_channel != 8)
        XimError("Iimage must have 24 bits per pixel (3 channels).");

    if (enhance)
        Edge_enhance(&xim, enhance);
    if (!(colors = (Color *)calloc((unsigned)ncolors, sizeof(Color))))
            XimError("Can't calloc() colors.");

    if (type == OCTREE) {
        if (!OctreeQuantize(&xim, ncolors))
            XimError("error in OctreeQuantize");
    } else if (type == MEDIAN) {
        if (map_flag)
            ncolors = MakeDithermap(ditherFactor, colors, ncolors);
        ncolors = XimQuant(&xim, ncolors, colors, map_flag, dither_flag);
    } else {
        if (ditherFactor < 2 || ditherFactor > 6)
            XimError("Dither level must be between 2 and 6");
        xim.ncolors = MakeDithermap(ditherFactor, colors, ncolors);
        Rgb_dither(&xim);
        if (!(xim.colors = (Color *)calloc((unsigned)xim.ncolors, 
                    sizeof(Color))))
            XimError("Can't calloc() colors.");
        for (i=0; i < xim.ncolors; i++) {
            xim.colors[i].red = colors[i].red;
            xim.colors[i].grn = colors[i].grn;
            xim.colors[i].blu = colors[i].blu;
        }
        if (debug_flag)
            for (i=0; i < ncolors; i++)
                fprintf(stderr, "%-3d:  (%3u, %3u, %3u) [%u]\n", i,
                colors[i].red, colors[i].grn, colors[i].blu,
                colors[i].pixel);
    }
    xim.bits_channel = 8;
    xim.nchannels = 1;
    RefineColors(&xim);
    WriteXim(out_fp, &xim);
    (void)fclose(out_fp);
    fprintf(stderr,"ncolors %d  nchannels %d\n", xim.ncolors,
            xim.nchannels);
    exit(0);
}


Usage ()
{
    static char **cpp, *help_message[] = {
    "where options include:",
    "  -in file           file to read in - stdin default",
    "  -out file          file to write out - stdout default",
    "  -dither[=#]        use dithering algorithm [variable levels]",
    "  -octree            use octree algorithm - default",
    "  -median            use median cut algorithm",
    "  -colors #          maximum number of colors in quantizing",
    "  -no #              number of pixels per node (octree only)",
    "  -cmap              use same colormap as dither (median only)",
    "  -enhance[=1-9]     variable edge enhancement, (default = 9)",
    "  -debug             print various debuging information",
    "Options may be abbreviated to the shortest unambiguous prefix.",
    };

    fprintf (stderr, "usage:\n  %s [-options ...]\n\n", progName);
    for (cpp = help_message; *cpp; cpp++)
        fprintf (stderr, "%s\n", *cpp);
    exit(1);
}

/*** end quantixim.c ***/
